container/state.go

```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
 	return true
 }
 
-func wait(waitChan <-chan struct{}, timeout time.Duration) error {
-	if timeout < 0 {
-		<-waitChan
-		return nil
-	}
-	select {
-	case <-time.After(timeout):
-		return fmt.Errorf("Timed out: %v", timeout)
-	case <-waitChan:
-		return nil
-	}
-}
+// WaitCondition is an enum type for different states to wait for.
+type WaitCondition int
+
+// Possible WaitCondition Values.
+//
+// WaitConditionNotRunning (default) is used to wait for any of the non-running
+// states: "created", "exited", "dead", "removing", or "removed".
+//
+// WaitConditionNextExit is used to wait for the next time the state changes
+// to a non-running state. If the state is currently "created" or "exited",
+// this would cause Wait() to block until either the container runs and exits
+// or is removed.
+//
+// WaitConditionRemoved is used to wait for the container to be removed.
+const (
+	WaitConditionNotRunning WaitCondition = iota
```

Comment:
What is the definition of "exited" though?.

The original implementation of `wait` considered a container which was in the `"created"` state to be *exited* (`wait` for a `"created"` container returns with exit status 0 immediately) but we explicitly want to block when the container has not yet started, both in the case of `docker run ...` and `docker start -a ...` where the container may either be created or even already exited and you are about to manually start it again. The `next-exit` condition is used in these two cases. You can think about the `next-exit` condition as a state transition, or as the equivalent of waiting for a `die` event for the container.